core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala

```diff
@@ -307,6 +331,71 @@ trait WorkspaceComponent {
       }
     }
 
+    def insertUserCatalogPermissions(workspaceId: UUID, userRefs: Seq[RawlsUserRef]) = {
+      val recordsToInsert = userRefs.map(userRef => WorkspaceUserCatalogRecord(workspaceId, userRef.userSubjectId.value))
+      (workspaceUserCatalogQuery ++= recordsToInsert)
+    }
+
+    def insertGroupCatalogPermissions(workspaceId: UUID, groupRefs: Seq[RawlsGroupRef]) = {
+      val recordsToInsert = groupRefs.map(groupRef => WorkspaceGroupCatalogRecord(workspaceId, groupRef.groupName.value))
+      (workspaceGroupCatalogQuery ++= recordsToInsert)
+    }
+
+    def deleteUserCatalogPermissions(workspaceId: UUID, userRefs: Seq[RawlsUserRef]) = {
+      val usersToRemove = userRefs.map(_.userSubjectId.value)
+      (workspaceUserCatalogQuery.filter(rec => rec.workspaceId === workspaceId && rec.userSubjectId.inSet(usersToRemove))).delete
+    }
+
+    def deleteGroupCatalogPermissions(workspaceId: UUID, groupRefs: Seq[RawlsGroupRef]) = {
+      val groupsToRemove = groupRefs.map(_.groupName.value)
+      (workspaceGroupCatalogQuery.filter(rec => rec.workspaceId === workspaceId && rec.groupName.inSet(groupsToRemove))).delete
+    }
+
+    def deleteWorkspaceCatalogPermissions(workspaceId: UUID) = {
+      workspaceUserCatalogQuery.filter(_.workspaceId === workspaceId).delete andThen
+        workspaceGroupCatalogQuery.filter(_.workspaceId === workspaceId).delete
+    }
+
+    def getUserCatalogPermissions(subjectId: RawlsUserSubjectId, workspaceContext: SlickWorkspaceContext): ReadAction[Boolean] = {
+      workspaceUserCatalogQuery.filter(rec => (rec.userSubjectId === subjectId.value && rec.workspaceId === workspaceContext.workspaceId)).countDistinct.result.map(rows => rows > 0) flatMap { hasCatalogPermission =>
+        if(hasCatalogPermission) DBIO.successful(hasCatalogPermission)
+        else rawlsGroupQuery.listGroupsForUser(RawlsUserRef(subjectId)).flatMap { userGroups =>
+          val groupNames = userGroups.map(_.groupName.value)
+          workspaceGroupCatalogQuery.filter(rec => ((rec.workspaceId === workspaceContext.workspaceId) && (rec.groupName).inSet(groupNames))).countDistinct.result.map(rows => rows > 0)
+        }
+      }
+    }
+
+    def listEmailsAndCatalog(workspaceContext: SlickWorkspaceContext): ReadAction[Seq[WorkspaceCatalog]] = {
+      val accessAndUserEmail = (for {
```

Comment:
These two `val`s duplicate the code in `listEmailsAndAccessLevel` - could you move them to new functions or methods so we don't duplicate?